### PR TITLE
Add function `enumerate_countfrom` to mirror Python's `enumerate(x, n)`

### PR DESCRIPTION
**Pull request by Tetralux**
~~Adds methods to `enumerate` so that we can specify a subset of an iterator that we wish to iterate over.~~

Addresses #19332.

Adds, and exports, function `enumerate_countfrom` which behaves exactly like `enumerate(iter, n)` in Python.

### CODE REVIEW

File: base/iterators.jl [hunk 1]
```diff
@@ -55,6 +55,18 @@ julia> for (index, value) in enumerate(a)
...
 enumerate(iter) = Enumerate(iter)
 
+"""
+    enumerate(iter, first_index)
+    enumerate(iter, first_index, last_index)
+    enumerate(iter, range)
+
+Enumerate a subset of an iterator.
+"""
+enumerate(iter, first::Integer) = enumerate(rest(iter, first))
```
Comment (martinholters) [inline base/iterators.jl]: `rest(iter, first)` only works if the state of `iter` equals the index. You probably need to use drop. Try e.g. `enumerate(filter(isodd, 1:10), ...)` for an example where using `rest` will fail.

File: test/iterators.jl [hunk 2]
```diff
@@ -391,3 +391,12 @@ let s = "Monkey 🙈🙊🙊"
     @test tf(2) == "Mo|nk|ey| 🙈|🙊🙊"
     @test tf(1) == "M|o|n|k|e|y| |🙈|🙊|🙊"
 end
+
+# Enumerating a subset of an iterator.
+let i = [1:5;]
+    @test collect(enumerate(i,4:5)) == [(1,4), (2,5)]
```
Comment (martinholters) [inline test/iterators.jl]: Try more interesting types for `i` here.

**Review by martinholters — DISMISSED** [review]

**Comment by ararslan** [comment]
I mentioned this in a [comment](https://github.com/JuliaLang/julia/issues/19332#issuecomment-260759743) on the original issue, but I think it would be nice to allow specifying both start and stop indices, ideally via keyword arguments. This can be accomplished using `take` and `drop` (Martin already mentioned the latter here).

File: base/iterators.jl (hunk 1 above)
Comment (H-225) [inline base/iterators.jl]: I don't think that example works, but I take your point :)

**Comment by H-225** [comment]
@ararslan 
> I mentioned this in a comment on the original issue, but I think it would be nice to allow specifying both  start and stop indices [...]

`first` and `last` _are_ the `start` and `stop` indices, are they not? - The first item in the result enumeration is the `first` one, and the last item is  the `last` one.

> [...] ideally via keyword arguments.

I have found that debugging keyword arguments is not very pleasant due to the machinery at play.
I'd rather that the stacktrace did not display things like:
`enumerate(iter; first::Int64=..., last::Int64=...)`
like 
`(::#kw##enumerate)(::Array{Any,1}, ::#enumerate, ::UnitRange{Int64})` 
before I use them in the Base.

**Comment by H-225** [comment]
@ararslan I added the method that takes a range, so that you can provide both values at the same time while maintaining clarity over what the values actually are.

**Comment by ararslan** [comment]
I like it, looks good. 👍 Nice work!

**Comment by ararslan** [comment]
Looks like you'll need to rebase and fix the branch conflicts.

**Comment by H-225** [comment]
Rebased.

**Comment by iamed2** [comment]
This might be better as a separate iterator. Python has `islice` (= "iterator slice") that returns an iterator over a region of an iterator.

**Comment by ararslan** [comment]
@iamed2 What would be the advantage of a separate iterator rather than separate methods for the existing iterator?

**Comment by iamed2** [comment]
You could combine it with all other possible iterator functions without adding methods to them.

**Comment by mbauman** [comment]
To my reading, this implements something different from what #19332 requested. Python's enumerate doesn't do this sort of subsetting — that's what `itertools.islice` does as @iamed2 notes.  Python's `enumerate(it, i)` still iterates over the entire iterator, but it changes where returned index counts from.  

```py
In [1]: for i, el in enumerate("Hello"): print(i, el)
0 H
1 e
2 l
3 l
4 o

In [2]: for i, el in enumerate("Hello", 10): print(i, el)
10 H
11 e
12 l
13 l
14 o
```

I think it'd be confusing to behave differently here.

**Comment by H-225** [comment]
@mbauman 
Having re-read the request, I would agree that is what they meant. However, I suspect that is not a good idea as it would not be clear to me what `enumerate(iter, 10)` does, without prior knowledge.
I'd rather have a separate function which has this behaviour - `enumerate_countfrom` is the clearest I've got so far.

@iamed2 
Something like, `for (i, e) in enumerate(slice(iter, 1:10))` could work. Bit verbose, but it works.

**Comment by TotalVerb** [comment]
Iterator slicing can (and should) be implemented in terms of take and drop.

**Comment by H-225** [comment]
@TotalVerb Do you mean that we _should not_ have a function for slicing iterators and just rely on calling `take(drop(...), ...)` directly?

**Comment by TotalVerb** [comment]
No, I mean that if we do have such a function, then its implementation should be in terms of the more elementary functions.

**Comment by H-225** [comment]
@TotalVerb You mean like, `slice(iter, ...) = take(drop(iter, ...), ...)` ?

**Comment by TotalVerb** [comment]
Yes, if we were to add `slice`. I'm not convinced it needs to be in Base.

**Comment by H-225** [comment]
I have removed all iterator slicing from this PR since that part is more controversial.
This PR now only addresses #19332 by adding `enumerate_countfrom(x, n::Integer)`.

**Comment by ararslan** [comment]
`enumerate_countfrom` is quite long. How about `enumeratefrom`?

**Comment by H-225** [comment]
@ararslan 
I did consider that one - `enumeratefrom` is too easy to confuse with "enumerate from the n-th item".
`enumerate_countfrom` is _way_ clearer, particularly since the usage will be `enumerate_countfrom(x, 10)`; "enumerate `x`, counting from `10`"
There's also the `countfrom` function, and since this is equivalent to `zip(countfrom(10), x)`...

**Comment by TotalVerb** [comment]
`zip(countfrom(10), x))` is actually shorter than `enumerate_countfrom(x, 10)`, and is one less function to learn, and is pretty readable :-/.

**Comment by martinholters** [comment]
`zip(countfrom(10), x)` also seems to be faster than the `enumerate_countfrom(x, 10)` proposed here.

**Comment by H-225** [comment]
@martinholters 
With the following code, my results are the opposite.
I would use `BenchmarkTools`, but that didn't work on my system.
```julia
julia> (@elapsed begin;
              r = 1:2000
              n = 1;
              for i in 1:1000;
                  for (j,e) in enumerate_countfrom(r, 10);
                     n += j;
                  end
              end
          end) / 1000
0.001409521346

julia> (@elapsed begin;
              r = 1:2000
              n = 1;
              for i in 1:1000;
                  for (j,e) in zip(Iterators.countfrom(10), r);
                     n += j;
                  end
              end
          end) / 1000
0.00238524494
```

@TotalVerb I don't disagree.

**Comment by TotalVerb** [comment]
Performance should be of no concern. One could always specialize `zip(::Count, xs)` if needed.

**Comment by StefanKarpinski** [comment]
`enumerate(itr, start)` seems like a reasonable enough addition to me.

**Pull request closed without merging**